lib/logstash/inputs/file.rb

```diff
@@ -184,44 +213,83 @@ def register
     @codec = LogStash::Codecs::IdentityMapCodec.new(@codec)
   end # def register
 
+  class ListenerTail
+    # use attr_reader to define noop methods
+    attr_reader :input, :path, :data
+    attr_reader :deleted, :created, :error, :eof
+
+    # construct with upstream state
+    def initialize(path, input)
+      @path, @input = path, input
+    end
+
+    def timed_out
+      input.codec.evict(path)
+    end
+
+    def accept(data)
+      # and push transient data filled dup listener downstream
+      input.log_line_received(path, data)
+      input.codec.accept(dup_adding_state(data))
+    end
+
+    def process_event(event)
+      event["[@metadata][path]"] = path
+      event["path"] = path if !event.include?("path")
+      input.post_process_this(event)
+    end
+
+    def add_state(data)
+      @data = data
+      self
+    end
+
+    private
+
+    # duplicate and add state for downstream
+    def dup_adding_state(line)
+      self.class.new(path, input).add_state(line)
+    end
+  end
+
+  def listener_for(path)
+    # path is the identity
+    ListenerTail.new(path, self)
+  end
+
   def begin_tailing
-    stop # if the pipeline restarts this input.
-    @tail = FileWatch::Tail.new(@tail_config)
+    # if the pipeline restarts this input.
+    stop
```

Comment:
The stop here is to close unwanted open fd so we don't leak them?
Might be good to add comment why we actually call stop here.